feat(g-code-testing): Add functionality to run HTTP requests through G-Code Parser (RET-148, RET-149, RET-151)

DerekMaggio opened the pull request:
# Overview

Add functionality to run HTTP requests through G-Code Parser 

# Changelog

* Refactor `ProtocolRunner` class into `GCodeEngine` ABC. Under which are the following classes
  * ProtocolGCodeEngine - For running protocol files and getting G-Code output
  * HTTPGCodeEngine - For running methods attached to HTTP requests and getting G-Code output
* Refactor `cli.py` to take a "configuration_name", when executing `run` command, instead of a path to a Python file
  * Did this to create an abstraction layer to run multiple input sources through G-Code Engine i.e.(HTTP and Protocols)
* Configurations are now defined in [configurations.json](https://github.com/Opentrons/opentrons/pull/8316/files#diff-10890de1fa4a92c85e11c31713a8e3d61e7589811ef822c2ddf9200bdf8fc1cf)
* Created [g_code_test_data.py](https://github.com/Opentrons/opentrons/pull/8316/files#diff-cad1a13f51d792f45d5225e191d08f23d9ac905b86ee1a7a77facd7e4333d852) to parse configuration.json
  * Added pydantic to g-code-testing for this purpose
* Added G-Code generation code for the following HTTP requests
  * `robot/move`
  * `robot/home`
* Updated README to account for new changes
* Updated Makefile to account for new changes
* Add tests for G-Code Engine expansion
* Add Github Action for running linting and tests against `g-code-testing` project

# Review requests

* Is there something better I can do than `asyncio.run` in g_code_engine and the configuration to actually run the async functions?
  * In configuration: https://github.com/Opentrons/opentrons/blob/2c4e55625cfdf638455f6da7f689e26af84bbf37/g-code-testing/g_code_test_data/http/robot/home/right_pipette.py#L17
  *  In Engine: https://github.com/Opentrons/opentrons/blob/2c4e55625cfdf638455f6da7f689e26af84bbf37/g-code-testing/g_code_parsing/g_code_engine.py#L160

# Risk assessment

None, it's new functionality

Commit:
Filter out __init__.py files

Commit:
feat(g-code-testing): Add robot HTTP support to G-Code Project

* Convert ProtocolRunner into an ABC called GCodeEngine
    * In ABC define an abstract `run` method
    * Rename all references
* Create ProtocolGCodeEngine which runs protocols
* Create HTTPGCodeEngine which runs HTTP requests

Commit:
Push name change

Commit:
Create first configuration

Commit:
Fix import paths in README

Commit:
Rename references in README

Commit:
Formatting

Commit:
Change folder name

Commit:
Turn into package

Commit:
Add new test data

Commit:
Add configuration file to define different test data

Commit:
Update cli.py to work with configuration file

Commit:
Fix JSON file

Commit:
Update Makefile

Commit:
Formatting

Commit:
Use hardcoded versions

Commit:
Fix tests

Commit:
Fix variable name

Commit:
Refactor from_config_file to add dependency injection for testing

* Also renamed classes to not start with `Test` because it was messing up pytest

Commit:
Add testing

Commit:
Formatting

Commit:
Linting

Commit:
Add Github Action

amitlissack commented on g-code-testing/g_code_parsing/cli.py:
```diff
@@ -151,7 +153,7 @@ def parse_args(cls, args: List[str]) -> Dict[str, Any]:
             'Concise',
             '--left-pipette',
-            '{"model": "p20_single_v2.0", "id": "P20SV202020070101"}',
+            '{"model": "p20_single_v2.0", "id": "P20SV202020070101 "}',
```
there's an extra space

amitlissack replied:
I think it's still there :)

DerekMaggio replied:
Thought I fixed it. 
Oh well, it's fixed now

amitlissack commented on g-code-testing/g_code_parsing/cli.py:
```diff
@@ -55,10 +54,20 @@ def execute(self) -> str:
+        if configuration.driver == ValidDriverTypes.PROTOCOL.value:
+            engine = ProtocolGCodeEngine(settings)
+
+        if configuration.driver == ValidDriverTypes.HTTP.value:
```
change to `elif` please

amitlissack commented on g-code-testing/g_code_parsing/g_code_engine.py:
```diff
@@ -102,24 +105,58 @@ def _get_protocol(file_path: str) -> Protocol:
+        # TODO: The below code is very implicit and I don't like it.
+        # It should get fixed
+        module = import_module(module_string)
+        return module.main()  # type: ignore
```
Will this raise if there's no main function?

DerekMaggio replied:
No, I am thinking about having a ABC class that requires a `main` function. All the HTTP cases would be a 
 class that needs to inherit from it.

I was thinking about registering all those functions somehow and having them be looked up by the configuration name instead of importing a random string.

amitlissack commented on g-code-testing/g_code_parsing/g_code_engine.py:
```diff
@@ -102,24 +105,58 @@ def _get_protocol(file_path: str) -> Protocol:
+        server_manager = ServerManager(self._config)
+        self._start_emulation_app(server_manager)
+        with GCodeWatcher() as watcher:
+            asyncio.run(func(hardware=self._emulate_hardware()))
```
While this is clever (not in a bad way!), I think it would be easier to understand if `main` accepted a `hardware` arg rather than returning a function that accepts a `hardware` arg.

DerekMaggio replied:
The question I have for that is, will `main` always need a `hardware` arg though?

amitlissack replied:
Can you think of an example where it wouldn't?

DerekMaggio replied:
Maybe something under robot_server/robot/calibration?
I haven't figured out how that portion of the code works yet.

amitlissack reviewed (commented):
I think this is a really smart and flexible solution. 

I'd like to see the spec of the HTTP script a bit simplified, if that's possible.

Commit:
Split up GCodeTestData

Break out GCodeTestData into HTTPTestData and ProtocolTestData. 
Did this so parameters on each class can be more specific than just path.

Commit:
Update GCodeEngine to account for TestData breakout

Instead of subclassing and having a single method just have more specialized methods on GCodeEngine. 

Pass HTTPTestData or ProtocolTestData classes itself instead of parameter from class

Commit:
Change from JSON to Py. Update config files

Changed json file to py file so for the HTTP types I can specify paths to the main function

Changed all HTTP configs to inherit from HTTPBase and defined a main function

Commit:
Forgot to ignore imports from robot_server

Commit:
Update tests

Commit:
Update to account for TestData breakout

amitlissack reviewed (approved):
👍

Commit:
Fix whitespace
